npm/pkg/dataplane/ipsets/ipset.go

```diff
@@ -0,0 +1,191 @@
+	"NameSpace":                1,
+	"KeyLabelOfNameSpace":      2,
+	"KeyValueLabelOfNameSpace": 3,
+	"KeyLabelOfPod":            4,
+	"KeyValueLabelOfPod":       5,
+	"NamedPorts":               6,
+	"NestedLabelOfPod":         7,
+	"CIDRBlocks":               8,
+}
+
+func (x SetType) String() string {
+	return SetTypeName[int32(x)]
+}
+
+func GetSetType(x string) SetType {
+	return SetType(SetTypeValue[x])
+}
+
+type SetKind string
+
+const (
+	ListSet SetKind = "list"
+	HashSet SetKind = "set"
+)
+
+func NewIPSet(name string, setType SetType) *IPSet {
+	set := &IPSet{
+		Name:              name,
+		HashedName:        util.GetHashedName(name),
+		Type:              setType,
+		SelectorReference: make(map[string]struct{}),
+		NetPolReference:   make(map[string]struct{}),
+		IpsetReferCount:   0,
+	}
+	if getSetKind(set) == HashSet {
+		set.IPPodKey = make(map[string]string)
+	} else {
+		set.MemberIPSets = make(map[string]*IPSet)
+	}
+	return set
+}
+
+func (set *IPSet) GetSetContents() ([]string, error) {
+	contents := make([]string, 0)
+	setType := getSetKind(set)
+	switch setType {
+	case HashSet:
+		for podIP := range set.IPPodKey {
+			contents = append(contents, podIP)
+		}
+		return contents, nil
+	case ListSet:
+		for _, memberSet := range set.MemberIPSets {
+			contents = append(contents, memberSet.HashedName)
+		}
+		return contents, nil
+	default:
+		return contents, fmt.Errorf("Unknown set type %s", setType)
+	}
+}
+
+func getSetKind(set *IPSet) SetKind {
+	switch set.Type {
+	case CIDRBlocks:
+		return HashSet
+	case NameSpace:
+		return HashSet
+	case NamedPorts:
+		return HashSet
+	case KeyLabelOfPod:
+		return HashSet
+	case KeyValueLabelOfPod:
+		return HashSet
+	case KeyLabelOfNameSpace:
+		return ListSet
+	case KeyValueLabelOfNameSpace:
+		return ListSet
+	case NestedLabelOfPod:
+		return ListSet
+	default:
+		return "unknown"
+	}
+}
+
+func (set *IPSet) AddMemberIPSet(memberIPSet *IPSet) {
+	set.MemberIPSets[memberIPSet.Name] = memberIPSet
+}
+
+func (set *IPSet) IncIpsetReferCount() {
+	set.IpsetReferCount++
+}
+
+func (set *IPSet) DecIpsetReferCount() {
+	set.IpsetReferCount--
+}
+
+func (set *IPSet) AddSelectorReference(netPolName string) {
+	set.SelectorReference[netPolName] = struct{}{}
+}
+
+func (set *IPSet) DeleteSelectorReference(netPolName string) {
+	delete(set.SelectorReference, netPolName)
+}
+
+func (set *IPSet) AddNetPolReference(netPolName string) {
+	set.NetPolReference[netPolName] = struct{}{}
+}
+
+func (set *IPSet) DeleteNetPolReference(netPolName string) {
+	delete(set.NetPolReference, netPolName)
+}
+
+func (set *IPSet) CanBeDeleted() bool {
+	if len(set.SelectorReference) > 0 {
+		return false
+	}
+	if len(set.NetPolReference) > 0 {
+		return false
+	}
+	if set.IpsetReferCount > 0 {
+		return false
+	}
+	if len(set.MemberIPSets) > 0 {
+		return false
+	}
+	if len(set.IPPodKey) > 0 {
+		return false
+	}
+	return true
+}
+
+func (set *IPSet) UsedByNetPol() bool {
+	if len(set.SelectorReference) <= 0 {
```

Comment:
length can never be negative, so maybe == 0?

Reply:
you should make the same suggestion as above! `return len(set.SelectorReference) > 0 || len(set.NetPolReference) > 0`